File: examples/play_sgf.cpp
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>

#include "go_data_gen/board.hpp"
#include "go_data_gen/sgf.hpp"

using namespace go_data_gen;

int main(int argc, char* argv[]) {
    if (argc != 2) {
        printf("Usage: %s <sgf_file_path>\n", argv[0]);
        return 1;
    }

    std::string file_path = argv[1];
    Board board;
    std::vector<Move> moves;
    float result;

    load_sgf(file_path, board, moves, result);

    printf("Board after setup:\n");
    board.print();

    int i = 1;
    for (const auto& move : moves) {
        printf("Playing move no. %d, which is: %s ", i,
               move.color == Color::Black ? "Black" : "White");
        if (move.is_pass) {
            printf("pass\n");
        } else {
            printf("(%d, %d)\n", move.coord.x, move.coord.y);
        }
        board.play(move);
        board.print();

        // Check after some moves
        if (i >= 180 && i <= 183 || i == 224) {
            printf("NN Input Data after %d moves:\n", i);
            const auto feature_planes = board.get_feature_planes(opposite(move.color));
            assert(feature_planes[0][0].size() == board.num_feature_planes);
            for (int c = 0; c < board.num_feature_planes; ++c) {
                printf("Feature plane %d: \n", c);
                board.print([&feature_planes, c](int x, int y) {
                    return static_cast<bool>(feature_planes[y][x][c]);
                });
            }
            board.print_group_sizes();
            board.print_liberties();
            const auto feature_vector = board.get_feature_scalars(opposite(move.color));
            assert(feature_vector.size() == board.num_feature_scalars);
            for (int c = 0; c < board.num_feature_scalars; ++c) {
                printf("Scalar feature %d = %f\n", c, feature_vector[c]);
            }
        }

        ++i;
    }

    printf("Result: ");
    if (result > 0) {
        printf("B+%.1f\n", result);
    } else if (result < 0) {
        printf("W+%.1f\n", -result);
    } else {
        printf("Draw or void game\n");
    }

    return 0;
}

File: examples/play_sgf.py
import sys
import go_data_gen


def main():
    if len(sys.argv) != 2:
        print(f"Usage: {sys.argv[0]} <sgf_file_path>")
        return 1

    file_path = sys.argv[1]
    is_valid, board, moves, result = go_data_gen.load_sgf(file_path)

    if not is_valid:
        print(f"Error: The SGF file {file_path} is not suitable for training.")
        return 1

    print("Board after setup:")
    board.print()

    for i, move in enumerate(moves, start=1):
        print(
            f"Playing move no. {i}, which is: {'Black' if move.color == go_data_gen.Color.Black else 'White'} ", end="")
        if move.is_pass:
            print("pass")
        else:
            print(f"({move.coord.x}, {move.coord.y})")

        board.play(move)
        board.print()

        # Check after specific moves
        if i >= 180 and i <= 183 or i == 224:
            print(f"NN Input Data after {i} moves:")
            feature_planes = board.get_feature_planes(
                go_data_gen.opposite(move.color))
            assert len(feature_planes[0][0]) == board.num_feature_planes

            for c in range(board.num_feature_planes):
                print(f"Feature plane {c}:")
                board.print(lambda x, y: bool(feature_planes[y][x][c]))

            board.print_group_sizes()
            board.print_liberties()

            feature_vector = board.get_feature_scalars(
                go_data_gen.opposite(move.color))
            assert len(feature_vector) == board.num_feature_scalars
            for c in range(board.num_feature_scalars):
                print(f"Scalar feature {c} = {feature_vector[c]}")

    print("Result: ", end="")
    if result > 0:
        print(f"B+{result:.1f}")
    elif result < 0:
        print(f"W+{-result:.1f}")
    else:
        print("Draw or void game")

    return 0


if __name__ == "__main__":
    sys.exit(main())
